### ISSUE TITLE
`File` should not be a subclass of `str`

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#799]
File behaves differently on python 3.5 vs 3.6

Python >= 3.6 introduces a protocol, `__fspath__`, which allows implementing objects to be passed as
a parameter to file operations such as `open`. The current File implementation implements that
protocol to return a posix path that refers to where Parsl expects that file to concretely exist
on the file system. In order to support this, there is no particular requirement on the superclass of File.

Python 3.5 does not support the `__fspath__` protocol, and only `str` or `byte` objects can be passed to file operations such as `open`.

To attempt to provide similar functionality, File subclasses `str`, which means it can be passed to `open`. It then overrides `__str__()` to provide a new string representation that is dynamically generated.

However, this does not always work: `open` uses the immutable string value used to construct the
`str` parent of File, not the value returned by `__str__`.

When a `File` is constructed like this: `File("./out.txt")` then the immutable str value is (probably) correct and no error occurs. The existing test cases catch this.

When a `File` is constructed like this: `File("file:///tmp/out.txt")` then the immuatable str value is that entire URI, even though invoking `__str__()` returns a reasonable posix path: `/tmp/out.txt`

This case of opening a URI-based File is not tested in the current test suite.

Because the immutable str value, and the result of `__str__()` do not agree, some awkward error messages arise. For example, an error message like this is reported:

```
Exception: [Errno 2] No such file or directory: /tmp/qwertz.out
```

even though /tmp/qwertz.out is somewhere writeable, because (according to `strace`) the OS open call was invoked with the immutable str value:

```
open("file:///tmp/qwertz.out", O_RDWR|O_CREAT|O_APPEND|O_CLOEXEC, 0666 <un
finished ...>
```

This bug only affects python 3.5, not python 3.6.

My recommendation is to remove the str superclass of File. This should not change behaviour on python >= 3.6, but would require a change in how File objects are using with Python 3.5: it would be necessary to explicitly convert the object into a path when passing it into `open`: for example, by calling `str(file)` explicitly.

**Comment by benclifford** [comment]
crossref #174 #175

**Comment by benclifford** [comment]
Fixed by #880
